### PR TITLE
fix(stripe): prevent duplicate subscription creation when a subscription already exists

### LINKED ISSUE #7099: Duplicate rows created in Subscription table on upgrade (Stripe)
### Is this suited for github?

- [x] Yes, this is suited for github

### To Reproduce

1. Create a User with no subscription
2. Upgrade to a paid subscription
3. After successful upgrade, two database records in the Subscription table are created, both completety identical with only the "id" being different.

### Current vs. Expected behavior

In 1.4.9 and below, when I upgraded a user from a free plan (no Subscription attached, no row in Subscription table) to a paid plan, only one row is created in the Subscription table for the user.

In 1.4.10, two identical rows are created after successful plan upgrade. Both are 100% identical, except the "id"

I am using Stripe local webhook listener:
stripe listen --forward-to localhost:3000/api/auth/stripe/webhook

This causes some issues, for example when I want to Cancel my subscription, only one of the rows will get its "cancelAt" date updated.

### What version of Better Auth are you using?

1.4.10

### System info

```bash
{
  "system": {
    "platform": "darwin",
    "arch": "arm64",
    "version": "Darwin Kernel Version 24.6.0: Mon Jul 14 11:30:40 PDT 2025; root:xnu-11417.140.69~1/RELEASE_ARM64_T6041",
    "release": "24.6.0",
    "cpuCount": 14,
    "cpuModel": "Apple M4 Pro",
    "totalMemory": "48.00 GB",
    "freeMemory": "0.09 GB"
  },
  "node": {
    "version": "v22.17.0",
    "env": "development"
  },
  "packageManager": {
    "name": "npm",
    "version": "11.6.4"
  },
  "frameworks": [
    {
      "name": "next",
      "version": "^16.1.1"
    },
    {
      "name": "react",
      "version": "^19.2.3"
    }
  ],
  "databases": [
    {
      "name": "@prisma/client",
      "version": "^7.2.0"
    }
  ],
  "betterAuth": {
    "version": "Unknown",
    "config": null,
    "error": "(0 , _smtp2goNodejs.default) is not a function"
  }
}
```

### Which area(s) are affected? (Select all that apply)

Backend

### Auth config (if applicable)

```typescript
import { betterAuth } from "better-auth";
import { prismaAdapter } from "better-auth/adapters/prisma";
import { nextCookies } from "better-auth/next-js";
import { admin } from "better-auth/plugins";
import { stripe } from "@better-auth/stripe";
import Stripe from "stripe";
import {
  sendEmail,
  getVerificationEmailHtml,
  getPasswordResetEmailHtml,
  getEmailChangeVerificationHtml,
  getDeleteAccountVerificationHtml,
  getSubscriptionStartedEmailHtml,
  getSubscriptionRenewedEmailHtml,
  getPaymentFailedEmailHtml,
} from "@/lib/email";
import APP_CONFIG from "@/lib/config";
import prisma from "@/lib/db";
import { deleteObject } from "@/lib/server/s3";
import { env } from "@/lib/validations/env";
import { formatCurrency, formatDateLong } from "@/lib/utils";
import { isExternalUrl } from "@/lib/file-utils";

const stripeClient = new Stripe(env.STRIPE_SECRET_KEY, {
  apiVersion: "2025-11-17.clover",
});

export const auth = betterAuth({
  database: prismaAdapter(prisma, {
    provider: "postgresql",
  }),
  session: {
    cookieCache: {
      enabled: true,
      maxAge: 5 * 60,
    },
  },
  rateLimit: {
    enabled: true,
  },
  emailAndPassword: {
    enabled: true,
    requireEmailVerification: true,
    sendResetPassword: async ({ user, url }) => {
      await sendEmail({
        to: user.email,
        subject: `Reset your ${APP_CONFIG.name} password`,
        html: getPasswordResetEmailHtml(user.name || "there", url),
      });
    },
  },
  emailVerification: {
    sendOnSignUp: true,
    autoSignInAfterVerification: false,
    sendVerificationEmail: async ({ user, url }) => {
      await sendEmail({
        to: user.email,
        subject: user.name
          ? `Verify your email, ${user.name}`
          : "Verify your email",
        html: getVerificationEmailHtml(user.name || "there", url),
      });
    },
  },
  user: {
    changeEmail: {
      enabled: true,
      sendChangeEmailVerification: async ({ user, newEmail, url }) => {
        await sendEmail({
          to: user.email,
          subject: "Approve your email change",
          html: getEmailChangeVerificationHtml(
            user.name || "there",
            newEmail,
            url,
          ),
        });
      },
      callbackURL: "/auth/sign-in?info=email-changed",
    },
    deleteUser: {
      enabled: true,
      sendDeleteAccountVerification: async ({ user, url }) => {
        await sendEmail({
          to: user.email,
          subject: "Confirm account deletion",
          html: getDeleteAccountVerificationHtml(user.name || "there", url),
        });
      },
      beforeDelete: async (user) => {
        await prisma.$transaction(async (tx) => {
          const activeSubscriptions = await tx.subscription.findMany({
            where: {
              referenceId: user.id,
              status: { in: ["active", "trialing", "past_due"] },
            },
          });

          if (activeSubscriptions.length > 0) {
            throw new Error(
              "Cannot delete user with active subscriptions. Please cancel all subscriptions first.",
            );
          }

          await tx.subscription.deleteMany({
            where: { referenceId: user.id },
          });

          await tx.verification.deleteMany({
            where: { value: user.id },
          });
        });

        if (user.image && !isExternalUrl(user.image)) {
          await Promise.all([
            deleteObject(
              `${env.NEXT_PUBLIC_APP_NAME}/avatars/full/${user.image}`,
            ),
            deleteObject(
              `${env.NEXT_PUBLIC_APP_NAME}/avatars/thumbnail/${user.image}`,
            ),
          ]).catch(() => {});
        }
      },
    },
  },
  socialProviders: {
    google: {
      clientId: env.GOOGLE_CLIENT_ID,
      clientSecret: env.GOOGLE_CLIENT_SECRET,
    },
  },
  plugins: [
    admin({
      bannedUserMessage:
        "Your account has been banned. Please contact support for assistance.",
    }),
    stripe({
      stripeClient,
      stripeWebhookSecret: env.STRIPE_WEBHOOK_SECRET,
      createCustomerOnSignUp: true,
      subscription: {
        enabled: true,
        plans: [
          {
            name: "plus",
            priceId: "price_xxx",
            annualDiscountPriceId: "price_xxx",
            limits: {
              projects: 10,
              storage: 25,
            },
          },
          {
            name: "pro",
            priceId: "price_xxx",
            annualDiscountPriceId: "price_xxx",
            limits: {
              projects: -1,
              storage: -1,
            },
          },
        ],
        getCheckoutSessionParams: async () => ({
          params: {
            tax_id_collection: {
              enabled: true,
            },
            billing_address_collection: "required",
            customer_update: {
              name: "auto",
              address: "auto",
            },
          },
        }),
        onSubscriptionComplete: async ({
          subscription,
          stripeSubscription,
          plan,
        }) => {
          try {
            const user = await prisma.user.findFirst({
              where: { stripeCustomerId: subscription.stripeCustomerId },
            });
            if (!user?.email) return;

            const amount = formatCurrency(
              stripeSubscription.items.data[0].price.unit_amount!,
            );
            const nextBillingDate = subscription.periodEnd
              ? formatDateLong(subscription.periodEnd)
              : "N/A";

            let invoiceUrl: string | undefined;
            if (stripeSubscription.latest_invoice) {
              const invoiceId =
                typeof stripeSubscription.latest_invoice === "string"
                  ? stripeSubscription.latest_invoice
                  : stripeSubscription.latest_invoice.id!;

              const invoice = await stripeClient.invoices.retrieve(invoiceId);
              invoiceUrl = invoice.hosted_invoice_url ?? undefined;
            }

            await sendEmail({
              to: user.email,
              subject: `Welcome! ${plan.name.charAt(0).toUpperCase() + plan.name.slice(1)} subscription active`,
              html: getSubscriptionStartedEmailHtml(
                user.name || "there",
                plan.name,
                amount,
                nextBillingDate,
                invoiceUrl,
              ),
            });
          } catch (error) {
            // Handle error
          }
        },
      },
      async onEvent(event) {
        try {
          if (!event.type.startsWith("invoice.")) return;

          const invoice = event.data.object as Stripe.Invoice;
          const customerId =
            typeof invoice.customer === "string"
              ? invoice.customer
              : invoice.customer?.id;
          if (!customerId) return;

          const user = await prisma.user.findFirst({
            where: { stripeCustomerId: customerId },
          });
          if (!user?.email) return;

          if (
            event.type === "invoice.payment_succeeded" &&
            invoice.billing_reason === "subscription_cycle"
          ) {
            const amount = formatCurrency(invoice.amount_paid ?? 0);
            const nextBillingDate = invoice.period_end
              ? formatDateLong(invoice.period_end)
              : "N/A";

            await sendEmail({
              to: user.email,
              subject: "Subscription renewed",
              html: getSubscriptionRenewedEmailHtml(
                user.name || "there",
                amount,
                nextBillingDate,
                invoice.hosted_invoice_url ?? undefined,
              ),
            });
          }

          if (event.type === "invoice.payment_failed") {
            const billingSession =
              await stripeClient.billingPortal.sessions.create({
                customer: customerId,
                return_url: `${env.BETTER_AUTH_URL}/dashboard/billing`,
              });

            await sendEmail({
              to: user.email,
              subject: "Payment failed - action required",
              html: getPaymentFailedEmailHtml(
                user.name || "there",
                formatCurrency(invoice.amount_due ?? 0),
                billingSession.url,
              ),
            });
          }
        } catch (error) {
          // Handle error
        }
      },
    }),
    nextCookies(),
  ],
  onAPIError: {
    errorURL: "/auth/error",
    onError: (error) => {
      // Handle error
    },
  },
  secret: env.BETTER_AUTH_SECRET,
  baseURL: env.BETTER_AUTH_URL,
});

export type Session = typeof auth.$Infer.Session;
```

### Additional context

Using the latest release of Better-Auth and Stripe 20.0.0

### PR DESCRIPTION
**Pull request by bytaesu**
Recently, `customer.subscription.created` event handler was added to handle subscription assignments from the Stripe dashboard, but the existing subscription check logic in that handler was weak. This PR fixes this issue.

- Closes https://github.com/better-auth/better-auth/issues/7099

---
## Summary by cubic
Prevent duplicate subscription creation in the Stripe customer.subscription.created webhook by checking metadata.subscriptionId before creating a new record. Ensures a single subscription per user and closes #7099.

- **Bug Fixes**
  - Update onSubscriptionCreated to look up by metadata.subscriptionId first, fallback to stripeSubscriptionId.
  - Add test verifying the webhook skips creating a duplicate when metadata.subscriptionId is present.

<sup>Written for commit c607d5ef6bee6f3bd18942fd5c0ea5728aed4f2f. Summary will update on new commits.</sup>

### CODE REVIEW

File: packages/stripe/src/hooks.ts [hunk 1]
```diff
@@ -138,19 +138,17 @@ export async function onSubscriptionCreated(
...
 				model: "subscription",
-				where: [
-					{
-						field: "stripeSubscriptionId",
-						value: subscriptionCreated.id,
-					},
-				],
+				where: subscriptionId
+					? [{ field: "id", value: subscriptionId }]
+					: [{ field: "stripeSubscriptionId", value: subscriptionCreated.id }], // Probably won't match since it's not set yet
```
[the remark this answers is not in the document]
Comment (bytaesu) [reply packages/stripe/src/hooks.ts:147]: In many cases, it doesn't exist, but it's included to handle it consistently with other webhooks.
